### PR TITLE
IOIENH-817 - errorHandling

### PR DESCRIPTION
**Pull request by vladanesic**

### CODE REVIEW

File: src/main/java/com/dais/ioi/external/domain/api/JmIntegrationApi.java [hunk 1]
```diff
@@ -78,8 +82,7 @@ CreateAccountResponse create( @RequestBody @Valid final CreateAccountRequest cre
                      method = RequestMethod.GET )
     @ApiOperation( value = "Get Policy Number" )
     GetPolicyNumberResponse getPolicyNumber( @RequestParam( value = "accountNumber" ) final String accountNumber,
-                                             @RequestParam( value = "jmSource",
-                                                            required = false ) final JmSource jmSource ); //TODO: make this field required when FE is ready
+                                             @RequestParam( value = "jmSource" ) final JmSource jmSource );
```
Comment (vladanesic) [inline src/main/java/com/dais/ioi/external/domain/api/JmIntegrationApi.java:85]: this is the fix from the previous task

File: src/main/java/com/dais/ioi/external/domain/api/JmIntegrationApi.java [hunk 2]
```diff
@@ -89,7 +92,7 @@ GetPolicyNumberResponse getPolicyNumber( @RequestParam( value = "accountNumber"
     List<UploadAppraisalResponse> uploadAppraisal( @RequestBody final UploadAppraisalRequestDto request );
 
     @ResponseStatus( HttpStatus.OK )
-    @RequestMapping( value = "/register/user/{lineId}",
+    @RequestMapping( value = "/register/user",
```
Comment (vladanesic) [inline src/main/java/com/dais/ioi/external/domain/api/JmIntegrationApi.java:95]: same as above

File: src/main/java/com/dais/ioi/external/controller/DefaultGlobalErrorAdvice.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+
+@Slf4j
+@ControllerAdvice
+public class DefaultGlobalErrorAdvice
+{
+
+    private static final String GENERIC_ERROR_RESPONSE = "We are unable to provide a response at this time.";
+
+    @Autowired
```
Comment (alicanbalikk) [inline src/main/java/com/dais/ioi/external/controller/DefaultGlobalErrorAdvice.java:23]: `@Autowired` is not a preferred way to inject dependencies, but l'd leave this choice to Edin or Aleksey.

I am bad at explaining things, but I'll try my best. I prefer constructor injection, because dependencies can be immutable for thread-safety. It does not allow you to initialise the service object without providing the required fields (properties) (This may sound nothing, but think of this way: A hacker somehow accesses the memory and injects `DefaultGlobalErrorAdvice` into their classes, and they can provide their `ObjectMapper` since it's auto-wired, or they don't provide anything and still manage to play with/use our service as they pleased. We don't want to allow others to inject their fields).

TL;DR - constructor injection is preferred approach instead of auto-wiring to prevent creating instances using default constructor. 

Check [this blog](https://odrotbohm.de/2013/11/why-field-injection-is-evil/) out for more info.

File: src/main/java/com/dais/ioi/external/controller/JmIntegrationController.java [hunk 4]
```diff
@@ -49,148 +47,63 @@
...
-            throw new ResponseStatusException( HttpStatus.INTERNAL_SERVER_ERROR, new String( e.content() ) );
-        }
+        log.info( "Received {}", objectMapper.writeValueAsString( firedTriggerDto ) );
 
-        catch ( Exception e )
-        {
-            log.error( e.getMessage(), e );
-            throw new ResponseStatusException( HttpStatus.INTERNAL_SERVER_ERROR, e.getMessage() );
-        }
+        triggerResponseDto = externalIntegrationService.process( firedTriggerDto );
```
Comment (alicanbalikk) [inline src/main/java/com/dais/ioi/external/controller/JmIntegrationController.java:57]:
```
log.info("Received {}", objectMapper.writeValueAsString(firedTriggerDto));

final TriggerResponseDto triggerResponseDto = externalIntegrationService.process(firedTriggerDto);
```

is cleaner than initialising `triggerResponseDto` with its default constructor and reinitializing it with a custom value.

File: src/main/java/com/dais/ioi/external/domain/dto/ErrorResponseDto.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+import java.util.List;
+
+
+@AllArgsConstructor
+@NoArgsConstructor
+@Data
+@Builder
+@JsonIgnoreProperties( ignoreUnknown = true )
+public class ErrorResponseDto
```
Comment (alicanbalikk) [inline src/main/java/com/dais/ioi/external/domain/dto/ErrorResponseDto.java:17]: I would prefer making this class immutable. `public final class ....`.

File: src/main/java/com/dais/ioi/external/controller/DefaultGlobalErrorAdvice.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+
+import java.util.Arrays;
+
+
+@Slf4j
+@ControllerAdvice
+public class DefaultGlobalErrorAdvice
```
Comment (alicanbalikk) [inline src/main/java/com/dais/ioi/external/controller/DefaultGlobalErrorAdvice.java:18]: I would prefer making this class immutable. `public final class ....`.

**Review by edoarsla — APPROVED** [review]
I have no context as to what the PR is wanting to accomplish.  But I cannot spot any errors.

File: src/main/java/com/dais/ioi/external/domain/dto/jm/UploadAppraisalRequestDto.java [hunk 7]
```diff
@@ -20,9 +21,12 @@
 @EqualsAndHashCode(callSuper = true)
 public class UploadAppraisalRequestDto extends JmBaseDto
 {
+    @NotNull
```
Comment (alicanbalikk) [inline src/main/java/com/dais/ioi/external/domain/dto/jm/UploadAppraisalRequestDto.java:24]: Properties of this dto now has constraints, but I don't see that the constraint checker `@Validated` is added to the request body of [uploadAppraisal( @RequestBody final UploadAppraisalRequestDto request )](https://github.com/dais-technology/ioi-external-integration/pull/173/files#diff-322dc44dbfce0bf5a5f9a6b49843c5cea5c8fec5159b7887544b4c01ed120756R92)

I believe it should be `uploadAppraisal( @RequestBody @Validated final UploadAppraisalRequestDto request )`

I may be wrong as I have not setup this project in my localhost, so I did not test this. If everything works well, ignore my comment.

File: src/main/java/com/dais/ioi/external/domain/dto/jm/UploadAppraisalRequestDto.java [hunk 8]
```diff
@@ -20,9 +21,12 @@
...
     private UUID clientId;
 
+    @NotNull
     private String accountNumber;
 
+    @NotNull
     private Map<String, ClientAnswerDto> intake;
```
Comment (alicanbalikk) [inline src/main/java/com/dais/ioi/external/domain/dto/jm/UploadAppraisalRequestDto.java:31]: In addition to my comment for line 24, since this is a collection whose value is a dto, `@Valid` annotation for `ClientAnswerDto` may be needed to trigger constraint checker for collection as it has a `@NotNull` field.

I may be wrong as I have not setup this project in my localhost, so I did not test this. If everything works well, ignore my comment.

**Review by alicanbalikk — COMMENTED** [review]
I have added my comments since I am assigned to review this PR, but since I am not a PR reviewer for back-end in general for DAIS, you can ignore my comments and continue as you pleased.
